packages/components/src/components/install-guide/ProjectPickerRequirements.vue

```diff
@@ -0,0 +1,275 @@
+import VNavigationButtons from '@/components/install-guide/NavigationButtons.vue';
+import Navigation from '@/components/mixins/navigation';
+import VQualityIcon from '@/components/QualityIcon.vue';
+
+export default {
+  name: 'project-picker-requirements',
+
+  components: {
+    VCodeSnippet,
+    VNavigationButtons,
+    VQualityIcon,
+  },
+
+  mixins: [Navigation],
+
+  props: {
+    project: {
+      type: Object,
+      default: () => {},
+    },
+    messageSuccess: {
+      type: String,
+      default: '<b>Copied!</b><br/>Paste this command<br/>into your terminal.',
+    },
+  },
+
+  computed: {
+    quality() {
+      if (!this.project) return undefined;
+      if (!this.project.score || this.project.score < 2) return 'bad';
+      if (this.project.score < 3) return 'ok';
+      return 'good';
+    },
+    installCommand() {
+      return [
+        'npx @appland/appmap install',
+        this.project && `-d ${this.project.path}`,
+      ]
+        .filter(Boolean)
+        .join(' ');
+    },
+    language() {
+      return (
+        (this.project && this.project.language && this.project.language.name) ||
+        'None Detected'
+      );
+    },
+    languageMessage() {
+      return (
+        (this.project && this.project.language && this.project.language.text) ||
+        ''
+      );
+    },
+    languageQuality() {
+      if (
+        !(this.project && this.project.language && this.project.language.score)
+      )
+        return 'bad';
+      if (this.project.language.score < 2) return 'bad';
+      if (this.project.language.score < 3) return 'ok';
+      return 'good';
+    },
+    webFramework() {
+      return (
+        (this.project &&
+          this.project.webFramework &&
+          this.project.webFramework.name) ||
+        'None Detected'
+      );
+    },
+    webFrameworkMessage() {
+      return (
+        (this.project &&
+          this.project.webFramework &&
+          this.project.webFramework.text) ||
+        ''
+      );
+    },
+    webFrameworkQuality() {
+      if (
+        !(
+          this.project &&
+          this.project.webFramework &&
+          this.project.webFramework.score
+        )
+      )
+        return 'bad';
+      if (this.project.webFramework.score < 2) return 'bad';
+      if (this.project.webFramework.score < 3) return 'ok';
+      return 'good';
+    },
+    testFramework() {
+      return (
+        (this.project &&
+          this.project.testFramework &&
+          this.project.testFramework.name) ||
+        'None Detected'
+      );
+    },
+    testFrameworkMessage() {
+      return (
+        (this.project &&
+          this.project.testFramework &&
+          this.project.testFramework.text) ||
+        ''
+      );
+    },
+    testFrameworkQuality() {
+      if (
+        !(
+          this.project &&
+          this.project.testFramework &&
+          this.project.testFramework.score
+        )
+      )
+        return 'bad';
```

Comment:
These guard conditionals are kind of gnarly, especially with Prettier formatting applied. My only advice would be to apply defaults to the potentially undefined values somewhere early on so you're guaranteed to always get a sane value.

Turning on optional chaining would also be beneficial, then we could boil this down to:
```js
if (!this.project?.testFramework?.score) return 'bad';
```

Something to consider for the future. I'm OK with leaving this as-is for now, it's pretty clear what's going on.

Reply:
Yup, optional chaining is much better, I just refactored it 😄